src/NuGet.Core/NuGet.Packaging/PackageExtraction/StreamExtensions.cs

```diff
@@ -27,11 +34,38 @@ public static string CopyToFile(this Stream inputStream, string fileFullPath)
                 return fileFullPath;
             }
 
+            // For files of a certain size, we can do some Cleverness and mmap
+            // them instead of writing directly to disk. This can improve
+            // performance by a lot on some operating systems and hardware,
+            // particularly Windows
+            long? size = null;
+            try
+            {
+                size = inputStream.Length;
+            }
+            catch (NotSupportedException)
+            {
+                // If we can't get Length, just move on.
+            }
             using (var outputStream = NuGetExtractionFileIO.CreateFile(fileFullPath))
             {
-                inputStream.CopyTo(outputStream);
+                if (size > 0 && size <= MAX_MMAP_SIZE)
+                {
+                    // NOTE: Linux can't create a mmf from outputStream, so we
+                    // need to close the file (which now has the desired
+                    // perms), and then re-open it as a memory-mapped file.
```

Comment:
This change was necessitated by Linux, but you've verified that it works well on all platforms?

Reply:
Tests are passing on all platforms, and performance improvements are holding on Windows. This is also an incredibly hot region as far as our tests are concerned. I only haven't manually checked on macOS, but we have our test suite running there (I suspect the single test failure is just a normal hiccup and doesn't have anything to do with these changes?)